frontend-plugin-core/src/main/java/com/github/eirslett/maven/plugins/frontend/lib/BunInstaller.java

```diff
@@ -0,0 +1,217 @@
+        this.config = config;
+        this.archiveExtractor = archiveExtractor;
+        this.fileDownloader = fileDownloader;
+    }
+
+    public BunInstaller setBunVersion(String bunVersion) {
+        this.bunVersion = bunVersion;
+        return this;
+    }
+
+    public BunInstaller setUserName(String userName) {
+        this.userName = userName;
+        return this;
+    }
+
+    public BunInstaller setPassword(String password) {
+        this.password = password;
+        return this;
+    }
+
+    public void install() throws InstallationException {
+        // use static lock object for a synchronized block
+        synchronized (LOCK) {
+            if (!bunIsAlreadyInstalled()) {
+                this.logger.info("Installing bun version {}", this.bunVersion);
+                if (!this.bunVersion.startsWith("v")) {
+                    this.logger.warn("Bun version does not start with naming convention 'v'.");
+                }
+                if (this.config.getPlatform().isWindows()) {
+                    throw new InstallationException("Unable to install bun on windows!");
+                } else {
+                    installBunDefault();
+                }
+            }
+        }
+    }
+
+    private boolean bunIsAlreadyInstalled() {
+        try {
+            BunExecutorConfig executorConfig = new InstallBunExecutorConfig(config);
+            File bunFile = executorConfig.getBunPath();
+            if (bunFile.exists()) {
+                final String version =
+                        new BunExecutor(executorConfig, Arrays.asList("--version"), null).executeAndGetResult(logger);
+
+                if (version.equals(this.bunVersion)) {
+                    this.logger.info("Bun {} is already installed.", version);
+                    return true;
+                } else {
+                    this.logger.info("Bun {} was installed, but we need version {}", version,
+                            this.bunVersion);
+                    return false;
+                }
+            } else {
+                return false;
+            }
+        } catch (ProcessExecutionException e) {
+            this.logger.warn("Unable to determine current bun version: {}", e.getMessage());
+            return false;
+        }
+    }
+
+    private void installBunDefault() throws InstallationException {
+        try {
+
+            logger.info("Installing Yarn version {}", bunVersion);
+
+
+            String downloadUrl = createDownloadUrl();
+
+            CacheDescriptor cacheDescriptor = new CacheDescriptor("bun", this.bunVersion,
+                    "zip");
+
+            File archive = this.config.getCacheResolver().resolve(cacheDescriptor);
+
+            downloadFileIfMissing(downloadUrl, archive, this.userName, this.password);
+
+            File installDirectory = getInstallDirectory();
+
+            // We need to delete the existing yarn directory first so we clean out any old files, and
```

Comment:
Same copy/paste typo problem as in some other places, couldn't the code be shared?

Reply:
fix with previous commit, thank you